
## LAB REPORT 1 Description of Assignment:

For this lab report, you will be working with the titanic ml dadataset. You would use this dataset to learn to wrangle dataframes.


## Data Description/Challenges:

The sinking of the Titanic is one of the most infamous shipwrecks in history.
On April $15$, $1912$, during her maiden voyage, the widely considered “unsinkable” RMS Titanic sank after colliding with an iceberg. Unfortunately, there weren’t enough lifeboats for everyone onboard, resulting in the death of $1502$ out of $2224$ passengers and crew.

In this challenge, you are tasked to do some wrangling on the data and find ways to understand the data a bit more. This aassignment is due Thursday, January 30, 2025


#### 1. import the necessary libraries

In [5]:
import pandas as pd

#### 1. import the necessary libraries

#### 2.  Load the train data in a variable name called "train"

In [29]:
train = pd.read_csv("/content/train.csv")

#### 3.  Load the test data in a variable name called "test"

In [8]:
test = pd.read_csv("/content/test.csv")

#### 4.  Check the shape of the train and test data

In [9]:
print("Shape of the test data:", test.shape)

Shape of the test data: (418, 11)


In [30]:
print("Shape of the train data:", train.shape)

Shape of the train data: (891, 12)


#### 5.  Comment on how many rows and columns the train and test data both have

In [31]:
#Rows and Columns of the train data
print("Rows of the train data:", train.shape[0])

#Columns of the train data
print("Columns of the train data:", train.shape[1])


Rows of the train data: 891
Columns of the train data: 12


In [32]:
#Rows and Columns of the test data
print("Rows of the test data:", test.shape[0])

#Columns of the test data
print("Columns of the test data:", test.shape[1])

Rows of the test data: 418
Columns of the test data: 11


The train data has 418 rows and 11 columns

The test data has 418 rows and 11 columns

#### 6.  Check the last 5 rows of the train and test data respectively.

In [33]:
#Checking the last 5 rows of the train data
train.tail(5)

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [34]:
#Checking the last 5 rows of the test data
test.tail(5)

PassengerId  Pclass                          Name     Sex   Age  SibSp  \
413         1305       3            Spector, Mr. Woolf    male   NaN      0   
414         1306       1  Oliva y Ocana, Dona. Fermina  female  39.0      0   
415         1307       3  Saether, Mr. Simon Sivertsen    male  38.5      0   
416         1308       3           Ware, Mr. Frederick    male   NaN      0   
417         1309       3      Peter, Master. Michael J    male   NaN      1   

     Parch              Ticket      Fare Cabin Embarked  
413      0           A.5. 3236    8.0500   NaN        S  
414      0            PC 17758  108.9000  C105        C  
415      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416      0              359309    8.0500   NaN        S  
417      1                2668   22.3583   NaN        C

#### 7.  Do a describitive of both the train and test data. Comment on the results from both data.

Both the last 5 rows of the train and test data have the same values for the PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked columns.

#### 8.  What are the column names in the train and test data respectively. Do you have any column in train which is absent in the test

The column names in the train data are PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin and Embarked.

The column names in the test data are PassengerId, Pclass, Name, Sex, Age, SibSp, Parch, Ticket, Fare, Cabin, Embarked.

No, there's no column in train which is absent in test.

#### 9.  Counts for the passengers who survived. Do this for only the train data

In [41]:
#Counts for the passengers who survived in the train data
train_count = (train['Survived']== 1).sum()
print(train_count)

#(df['A'] == 1).sum()

342


#### 10.  Extract only the titles from the name columns in the train
**Hint**: Check on using regular expression to extract the title from the name column.

In [42]:
#Extracting only the titles from the name columns in the train data
#titles = ['Mlle.', 'Major.', 'Master.' ,'Col.', 'Sir.', 'Don.', 'Mme.', 'Jonkheer.', 'Lady.', 'Capt.', 'Countess.', 'Ms.', 'Dona.']
#train_titles_extracted = train[train['Name'].str.contains('|'.join(titles), case=False, na=False)]
#print(train_titles_extracted)

train['Title'] = train['Name'].str.extract(r'(Mr|Mrs|Miss|Ms|Dr|Rev|Master)\.')
print(train[['Name', 'Title']])

                                                  Name Title
0                              Braund, Mr. Owen Harris    Mr
1    Cumings, Mrs. John Bradley (Florence Briggs Th...   Mrs
2                               Heikkinen, Miss. Laina  Miss
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)   Mrs
4                             Allen, Mr. William Henry    Mr
..                                                 ...   ...
886                              Montvila, Rev. Juozas   Rev
887                       Graham, Miss. Margaret Edith  Miss
888           Johnston, Miss. Catherine Helen "Carrie"  Miss
889                              Behr, Mr. Karl Howell    Mr
890                                Dooley, Mr. Patrick    Mr

[891 rows x 2 columns]


**Note:** You can only do this question after doing the previous question
 #### 11. Create a new column and call it "Title" to save the extracted titles above

In [44]:
#Saving the titles above into the "Title" column
# Regular expression to extract titles from the Name column
#train['Title'] = train['Name'].str.extract(r'(\b(?:Mlle.|Major.|Master.|Col.|Sir.|Don.|Mme.|Jonkheer.|Lady.|Capt.|Countess.|Ms.|Dona.)\b)', expand=True)
#'Mlle', 'Major', 'Master' ,'Col', 'Sir', 'Don', 'Mme', 'Jonkheer', 'Lady', 'Capt', 'Countess', 'Ms', 'Dona'

# Displaying the first few rows to verify the new Title column
#print(train[['Name', 'Title']].head(20))

train['Title'] = train['Name'].str.extract(r'(Mr|Mrs|Miss|Ms|Dr|Rev|Master)\.')
print(train)


     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

#### 12.  Use this dictionary to change the title names in the train.Title created column

mapping = {'Mlle': 'Miss', 'Major': 'Mr', 'Col': 'Mr', 'Sir': 'Mr',
           'Don': 'Mr', 'Mme': 'Mrs', 'Jonkheer': 'Mr', 'Lady': 'Mrs',
           'Capt': 'Mr', 'Countess': 'Mrs', 'Ms': 'Miss', 'Dona': 'Mrs'}

In [45]:
#Changing the title names in the train in the title created column
mapping = {'Mlle': 'Miss', 'Major': 'Mr', 'Col': 'Mr', 'Sir': 'Mr', 'Don': 'Mr', 'Mme': 'Mrs', 'Jonkheer': 'Mr', 'Lady': 'Mrs', 'Capt': 'Mr', 'Countess': 'Mrs', 'Ms': 'Miss', 'Dona': 'Mrs'}

train['Title'] = train['Title'].replace(mapping)
print(train[['Name', 'Title']].head())

                                                Name Title
0                            Braund, Mr. Owen Harris    Mr
1  Cumings, Mrs. John Bradley (Florence Briggs Th...   Mrs
2                             Heikkinen, Miss. Laina  Miss
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)   Mrs
4                           Allen, Mr. William Henry    Mr


#### 13.  Value counts after changing some titles

In [46]:
#The new value counts after changing some titles
value_count = train['Title'].value_counts()
print(value_count)

Title
Mr        517
Miss      183
Mrs       125
Master     40
Dr          7
Rev         6
Name: count, dtype: int64


#### 14. Group data by class and view the averages for each column

In [47]:
class_averages = train.groupby('Pclass').mean(numeric_only=True)
print(class_averages)

        PassengerId  Survived        Age     SibSp     Parch       Fare
Pclass                                                                 
1        461.597222  0.629630  38.233441  0.416667  0.356481  84.154687
2        445.956522  0.472826  29.877630  0.402174  0.380435  20.662183
3        439.154786  0.242363  25.140620  0.615071  0.393075  13.675550


#### 15.  What do you observe from the  above groupby table? Discuss and write down your thoughts

I observed that the Titanic dataset contains three classes:
1st class: Wealthier passengers.
2nd class: Middle-class passengers.
3rd class: Lower-class passengers or those with the cheapest tickets.

Under normal circumstances, the average of each column should have been a single value but there are three average values in each table which means that the titanic dataset contains three classes as described above.


#### 16.  Group by Pclass and sex and find the mean


In [48]:
class_and_sex_averages = train.groupby(['Pclass', 'Sex']).mean(numeric_only=True)
print(class_and_sex_averages)

               PassengerId  Survived        Age     SibSp     Parch  \
Pclass Sex                                                            
1      female   469.212766  0.968085  34.611765  0.553191  0.457447   
       male     455.729508  0.368852  41.281386  0.311475  0.278689   
2      female   443.105263  0.921053  28.722973  0.486842  0.605263   
       male     447.962963  0.157407  30.740707  0.342593  0.222222   
3      female   399.729167  0.500000  21.750000  0.895833  0.798611   
       male     455.515850  0.135447  26.507589  0.498559  0.224784   

                     Fare  
Pclass Sex                 
1      female  106.125798  
       male     67.226127  
2      female   21.970121  
       male     19.741782  
3      female   16.118810  
       male     12.661633  


#### 17. **Missing Values**

#### 18.  How many missing values are in train data  and test respectively.


In [49]:
#Missing values in the train data
train_missing_values = train.isnull().sum()
print(train_missing_values)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
Title           13
dtype: int64


In [50]:
#Missing values in the test data
test_missing_values = test.isnull().sum()
print(test_missing_values)

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64


#### 19.  Which of columns in the train data had missing values?

The columns in the train data that had missing values are Age, Embarked, Cabin, Title.

#### 20.  Which of columns in the test data had missing values?


The columns in the test data that had missing values are Age, Fare, Cabin.

#### 21.  First drop all missing values and check the counts for missing value again

In [51]:
#Dropping all missing values
train_values_dropped = train.dropna()
print(train_values_dropped)

     PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ... 

In [52]:
#Checking the counts for the missing values
train_values_recounted = train_values_dropped.isnull().sum()
print(train_values_dropped)

     PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ... 

**Remember** Do not run the code on dropping all missing values the second time

#### 22.  Reload the data and drop the missing values in age with the mean


In [53]:
#Reloading the data and dropping the missing values in age with the mean
import pandas as pd

# Reload the data
train = pd.read_csv("/content/train.csv")

# Calculate the mean of 'Age'
age_mean = train['Age'].mean()

# Replace missing values in 'Age' with the mean
train['Age'] = train['Age'].fillna(age_mean)

# Verify that missing values in 'Age' have been handled
print(f"Missing values in 'Age': {train['Age'].isnull().sum()}")


Missing values in 'Age': 0


#### 23. Drop the missing values with the Embarked with the mode


In [54]:
import pandas as pd

# Reload the data (replace the path with your actual file path)
train = pd.read_csv("/content/train.csv")

# Find the mode of the 'Embarked' column
embarked_mode = train['Embarked'].mode()[0]  # Get the most frequent value

# Replace missing values in 'Embarked' with the mode
train['Embarked'] = train['Embarked'].fillna(embarked_mode)

# Verify that missing values in 'Embarked' have been handled
print(f"Missing values in 'Embarked': {train['Embarked'].isnull().sum()}")


Missing values in 'Embarked': 0


#### 24. Merge the two dataframes and called it df.Check to the number of rows and columns and number of missing values

In [62]:
#import pandas as pd
#df = pd.read_csv("file path of the csv file")
#

import pandas as pd

# Assuming you have two DataFrames: df1 and df2
# Example:
# df1 = pd.DataFrame({'A': [1, 2], 'B': [3, 4]})
# df2 = pd.DataFrame({'A': [1, 2], 'C': [5, 6]})

# 1. Merge the two DataFrames (use 'on' or specify the type of merge if needed)


#Converting both Columns to strings
train['Age'] = train['Age'].astype(str)
test['Embarked'] = test['Embarked'].astype(str)

df = pd.merge(train, test, how='inner', left_on='Age', right_on='Embarked')  # You can change 'how' to 'outer', 'left', or 'right'

# 2. Check the number of rows and columns
print(f"Number of rows: {df.shape[0]}")
print(f"Number of columns: {df.shape[1]}")

# 3. Check for missing values
missing_values = df.isnull().sum().sum()  # Total missing values in the entire DataFrame
print(f"Number of missing values: {missing_values}")


Number of rows: 0
Number of columns: 23
Number of missing values: 0
